Articles/HowTos/CriticalBeginnerGitUsageTips.md

```diff
@@ -0,0 +1,99 @@
+Commit messages should have a short summary line (under 50 chars is the target) and an optional longer body separated by a blank line (see guidelines in [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/)).
+For example:
+
+```
+Summary line for changes (50 chars is target but a little longer okay)
+
+More detailed explanatory text in the body (wrapped to about 72 chars).
+The blank line separating the summary from the body is critical to make
+commands like 'git log --oneline', 'git rebase -i', and others readable
+and useful.
+
+If you use an issue tracker, put references to them like:
+
+Resolves: #123
+Also see: #456, #789
+```
+
+**C) Create small local "checkpoint" commits then cleanup with 'git rebase -i'
+before pushing.**
+To provide for easy local "undos" and better organization of changes into final change-sets, commit often locally using "checkpoint" commits (convention is to use commit summary lines starting with "WIP:").
+But before pushing these commits to a remote shared branch, use [`git rebase -i @{u}`](https://www.atlassian.com/git/tutorials/rewriting-history#git-rebase-i) to clean up and reorganize the commits into good "logical" commits (see "SEPARATE CHANGES" above).
+
+**D) Always create local commits for your changes in local branches before you run any Git commands that might modify your local uncommitted changes.**
+The commands `git pull`, `git merge`, `git rebase`, or other Git operations can alter (or
+delete) your local uncommitted changes, either in the working directory or the staging
+area/index.
+So always create (sometimes temporary) commits for these before running any of these commands (unless you want to throw away uncommitted changes in your local repo, for example with `git checkout`, `git reset`, `git clean`, etc.).
+
+**E) Back up local branches every few hours of work by pushing them to some
+remote Git repo on a remote machine** in case your local machine or disk goes out.
+(This is why local Git branches are better than Git stashes because you can back them up to other repos in a version controlled way.
+You can’t do that with Git stashes.)
+
+**F) You can always recover an earlier state of any of your local branches** by
+looking at [`git reflog`](https://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery) and then using a combination [`git checkout`](http://marklodato.github.io/visual-git-guide/index-en.html#checkout) and/or [`git reset –hard`](http://marklodato.github.io/visual-git-guide/index-en.html#reset), etc.
+(See [How to undo (almost) anything in Git](https://github.blog/2015-06-08-how-to-undo-almost-anything-with-git/#redo-after-undo-local).)
+
+**G) Never delete a local Git repo (unless you are completely done with it).**
```

Comment:
As an aside, I've done this exact thing (delete a local repo and re-clone) *many* times when I was working on Spack. The churn was so high, I couldn't keep up. And, the Git expertise necessary to unravel myself from it was more work than I could bear.

Reply:
FYI: I figured out how to clean a Spack repo so that you don't need to keep deleting and recloning it.  I wish Spack had a function to do this so you did not have to figure it out yourself.  Any tool that requires you to keep cloning and deleting a git repo (that you may have local changes in) is a poorly designed tool.

But that is a Spack issue and not a Git issue.  There is nothing wrong with Git.  That is a Spack problem (or any project that is sloppy about writing files all over the git source tree).